crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1438,6 +1441,32 @@ impl<'db> ClassType<'db> {
             })
     }
 
+    pub fn is_subclass_of(self, db: &'db dyn Db, other: Type) -> bool {
```

Comment:
Sorry I wasn't more clear about this: I think this method should take `other: ClassType` instead of `other: Type`, and attempt to only answer the much narrower question, "given two ClassType, does the one inherit the other". This would avoid a lot of the TODO questions below; I think more abstracted methods calling this should be responsible for handling e.g. unions and intersections. (For example, that is already the case in this PR, and would also be the case when e.g. `is_subtype_of` wants to make use of this method -- which it probably should already in this PR.)

That just leaves the question of how to handle `Any` in the MRO. I think for now (until we have a proper MRO and a clarified data model for what types can occur in the MRO) this method can just return `true` if it finds `Any/Unknown/Todo` or a matching `ClassType` in bases, otherwise `false`.

I think a doc comment clarifying the semantics of this method would be useful.
```suggestion
    /// Does this class inherit the given `other` class?
    pub fn is_subclass_of(self, db: &'db dyn Db, other: ClassType) -> bool {
```

Reply:
Since this PR is a joint work anyway, I went ahead and pushed what I was thinking here, so the review ball is now in your court :)

Reply:
Ah yes, you did indeed lead me astray here from your comment in https://github.com/astral-sh/ruff/pull/13800#discussion_r1806833944, when you said this 😄

> I think that the hypothetical `ClassType::is_subclass_of` should handle the special case of `Any` or `Unknown` in the MRO. (And unions too, if we support them in bases, though this might be handled via checking in multiple MROs instead.)

but no harm done!

Reply:
I think there was a misunderstanding between "handle Any/union in the MRO", which is what I said, and "handle Any/Union" in the `other` type", which is what you implemented. I didn't mean that comment to imply handling anything else in `other`. I don't think the version you implemented did handle Any or Union found in the MRO?

But my comment was wrong regardless, because in the end I realized I don't think we should handle Any in the MRO here.

Reply:
ah, yeah, good points all round!